include/tvm/diagnostic/internal/probe.h

```diff
@@ -0,0 +1,76 @@
+/** Copyright 2017-2021 CNRS-AIST JRL and CNRS-UM LIRMM */
+
+#pragma once
+
+#include <functional>
+#include <typeindex>
+#include <utility>
+
+namespace tvm::diagnostic::internal
+{
+/** Combine the hash value of \p v with the value \p seed
+ *
+ * Taken from https://stackoverflow.com/a/23860042
+ */
+template<typename T>
+inline void hash_combine(std::size_t & seed, const T & v) noexcept
+{
+  std::hash<T> hasher;
+  seed ^= hasher(v) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
+}
+
+/** Functor to hash a pair of object.
+ *
+ * Adapted from https://stackoverflow.com/a/23860042
+ */
+struct PairHasher
+{
+  template<typename T, typename U>
+  std::size_t operator()(const std::pair<T, U> & p) const noexcept
+  {
+    std::size_t hash = 0;
+    hash_combine(hash, p.first);
+    hash_combine(hash, p.second);
+    return hash;
+  }
+};
+
+/** An helper function to remove a noexcept from the signature of a method.
+ *
+ * Useful for the registration of methods in GraphProbe::registerAccessor
+ */
+template<typename T, typename U, typename... Args>
+auto remove_noexcept(U (T::*fn)(Args...) const noexcept)
+{
+  return static_cast<U (T::*)(Args...) const>(fn);
+}
```

Comment:
I had to use that to register some methods of `mc_rbdyn::Frame` in another project:
```
using T = mc_rbdyn::Frame;
gp.registerAccessor<T, sva::PTransformd>(T::Output::Position, tvm::diagnostic::internal::remove_noexcept(&T::position), [](const auto& H) {return Eigen::MatrixXd(H.matrix()); });
```
so I though it would be better as a utility here.

As for the legality, what would be the problem ? I'm removing a guarantee from the function. If it is `noexcept`, not having the guarantee is also true.

Reply:
> As for the legality, what would be the problem ? I'm removing a guarantee from the function. If it is noexcept, not having the guarantee is also true.

In C++17, `noexcept` is part of the function signature ([ref](https://en.cppreference.com/w/cpp/language/noexcept_spec)) and it seemed fishy at first but after looking more into it, maybe it's ok?

Regardless of legality, why not support it in `GraphProbe::registerAccessor` directly rather than having external users call into internal helpers that are not used internally?

To keep it relatively simple we could have something like:
```cpp
template<typename T, typename U, typename EnumOutput>
inline void GraphProbe::registerAccessor(EnumOutput o, U (Base::*fn)() const, std::function<Eigen::MatrixXd>(const U &)> convert)
{
  return registerAccessorImpl(o, fn, convert);
}

template<typename T, typename U, typename EnumOutput>
inline void GraphProbe::registerAccessor(EnumOutput o, U (Base::*fn)() const noexcept, std::function<Eigen::MatrixXd>(const U &)> convert)
{
  return registerAccessorImpl(o, fn, convert);
}


template<typename T, typename AccesorT, typename U, typename EnumOutput>
inline void GraphProbe::registerAccessorImpl(EnumOutput o, AccesorT fn, std::function<Eigen::MatrixXd>(const U &)> convert)
{
  /** Current implementation */
}
```

I've also investigated the option to use a single function to handle all cases with a dash of constexpr magic ([here](https://github.com/gergondet/tvm/commit/9dd665b2c0d3d746cc8bae0f9f77951158368dda))

With this proposal, the example you cited becomes:

```
using T = mc_rbdyn::Frame;
gp.registerAccessor<T>(T::Output::Position, &T::Position, [](const auto& H) {return Eigen::MatrixXd(H.matrix()); });
```

Note: for maximum performance we could also take the method pointer in the template arguments (we can do `template<auto Method> auto fun()` instead of `template<typename Method> auto fun(Method m)` but the cost of the indirection is negligible for something that is only meant for diagnostics and the syntax becomes a little ugly, as in `gp.register<T, &T::MyMethod>(T::Output::MyOutput, convert);`

Reply:
I took your proposed implementation. Thx :)